Basic Taproot signing support for descriptor wallets

sipa opened the pull request:
Builds on top of #22051, adding signing support after derivation support.

Nothing is changed in descriptor features. Signing works for key path and script path spending, through the normal sending functions, and PSBT-based RPCs. However, PSBT usability is rather low as no extensions have been defined to convey Taproot-specific information, so all script information must be known to the signing wallet.

sipa committed:
Add CKey::SignSchnorr function for BIP 340/341 signing

sipa committed:
Add precomputed txdata support to MutableTransactionSignatureCreator

This provides a means to pass in a PrecomputedTransactionData object to
the MutableTransactionSignatureCreator, allowing the prevout data to be
passed into the signature hashers. It is also more efficient.

sipa committed:
Permit full precomputation in PrecomputedTransactionData

At verification time, the to be precomputed data can be inferred from
the transaction itself. For signing, the necessary witnesses don't
exist yet, so just permit precomputing everything in that case.

sipa committed:
Don't nuke witness data when signing fails

sipa committed:
Construct and use PrecomputedTransactionData in SignTransaction

sipa committed:
Basic Taproot signing logic in script/sign.cpp

sipa committed:
tests: check spending of P2TR

sipa committed:
Add TaprootSpendData data structure, equivalent to script map for P2[W]SH

This data structures stores all information necessary for spending a taproot
output (the internal key, the Merkle root, and the control blocks for every
script leaf).

It is added to signing providers, and populated by the tr() descriptor.

sipa committed:
Construct and use PrecomputedTransactionData in PSBT signing

sipa committed:
Add support for SIGHASH_DEFAULT in RPCs, and make it default

For non-Taproot signatures, this is interpreted as SIGHASH_ALL.

sipa commented:
Rebased on top of #21246, renamed "inner key" everywhere in this PR to "internal key" as well, and added very basic PSBT support (you can have a watchonly tr(KEY) wallet now, and get it signed by an offline version that has the private key).

Sjors commented:
Concept ACK 🎉

roconnor-blockstream commented:
TBH, I find your "must be specified in depth-first search order" somewhat confusing.  As an alternative syntax I suggest simply using some notation for a binary or.  For example.

*  `tr(KEY)`: only taproot output key known (or is it an internal pubkey with common NUMS tweak??)
*  `tr(KEY,A)`: internal pubkey with a single script `A` at level 0.
*  `tr(KEY,[A|B])`: internal pubkey with a two scripts, `A` and `B` at level 1.
*  `tr(KEY.[[A|B]|C])`: internal pubkey with a two scripts, `A` and `B` at level 2, and a third script at level 1.
*  `tr(KEY.[A|[B|C]])`: same as above except for `A` is at level 1 and `C` is at level 2 instead.
*  `tr(KEY.[[B|C]|A])`: equal to the above descriptor.
*  `tr(KEY.[[A|B]|[C|D]])`: internal pubkey with a four scripts, all at level 2.
*  `tr(KEY.[[A|C]|[B|D]])`: same four scripts as above, again all at level 2 but now associated differently which makes for a different tweak.
*  `tr(KEY.[A|[B|[C|D]]])`: same four script as above, but no longer all at the same level.

My claimed advantages are the implicitness of the depth, and clearer notation indicating how the scripts are associated amongst each other.

The notation above is just an example and don't care too much about the specific syntax.

---

Other things to include in the list of features to be added later, a MerkleRoot descriptor for fixed but unknown branches.

sipa commented:
@roconnor-blockstream Yeah, I considered something like that, and am open to it. Both notations are equivalent in that you can fairly easily convert in both directions between them (counting bracing level in yours corresponds to the depth specified explicitly in mine, with leaves ultimately listed in the same order). The advantage I see to the current one is that I think it's a bit more readable, avoiding the need to go count braces to see relative depth of leaves. Yours is perhaps a bit easier to write.

There are many more variations possible. One extreme would be permitting arbitrary weights for every node, and have the descriptor expansion convert it to a Huffman tree. I'd prefer not to do this because it's adding information that can't be inferred back from the script tree it is converted to.

roconnor-blockstream commented:
Right.  My concern with explicit levels is that it is so easy to violate the invariant that the sum of the weights has to add up to 1, and you cannot really splice together subexpressions to compose a descriptor from fragments without going through and globally adjusting all the levels to be sane.

roconnor-blockstream commented:
Additionally I understand that it is the case that checking that the weights add up to 1 is not a sufficient condition to check that a string is a well-formed descriptor as it doesn't imply that the arrangement is in some DFS order. Thus I argue that my proposed scheme is better because the well-formedness of the descriptor format is a relatively simple syntactic check (that would pass [LANGSEC](https://langsec.org/) scrutiny).

JeremyRubin commented:
+1 @roconnor-blockstream  a bracketed version seems superior to me.

Some BIP-level (assuming this would eventually become a BIP) thoughts on things that might be useful in a descriptor:

1. it might make sense to have a node that's a Need-To-Know-Basis Subtree. N2KBS have some danger -- e.g., could be hiding a backdoor key -- but also can be used to simplify a descriptor for e.g. inter-operating with an Escrow service that you don't want to share the whole program with.
2. I think with brackets it's also easier to attach a subtree modifier. E.g., we could spec the lexing to recognize `[a-zA-Z0-9_]*(` or something similar as a subtree modifier. E.g., `[A,B]` -> `huffman[A@p,B@q]`. 

The advantage of having this baked in is great for readability -- often times you might want to express a tree as:

[Thing that should happen most likely, in the unlikely event that doesn't happen, Other thing that could happen, other thing we might want to do in unlikely event]

which would -- if depth order is required -- look like:

 [Thing that should happen most likely, Other thing that could happen, in the unlikely event that doesn't happen, other thing we might want to do in unlikely event]

Which reminds me of exception throwing where the error handling source code can be very far away from the source of the error (bad). 

(something like huffman does require that we also use a sorted Tree to select clauses lexicographically after probability).
3. An annotation/tag interface that would allow you to do something like `[A#{escrow, alice, bob}, B#{alice, bob}]` which when 'blind_for("escrow")' is called creates the `[A, N2KB]` descriptor.
4. Maybe a more general way of inserting extended structured metadata that Core will store/pass along but might not understand for supporting things like 3...
5. a `opaque[]` annotation, which compiles a subtree opaquely to any parent modifiers. This would be useful in contexts where a subtree might be known entirely already by another party and we want to preserve that structure

sipa commented:
@JeremyRubin I agree that there is a potential use for omitted subtrees, but I don't think there is much of a difference between the two approaches there? In the current approach you'd just specify the Merkle root of the omitted subtree as a "leaf" at the depth that root exists at.

All the other things you mention seem to be things for a policy language that get compiled to descriptors rather than descriptors themselves. Descriptors are for encoding information potentially needed by signers; I can't imagine a need for information of that kind that applies a specific subtree (rather than applying to individual leaves, or to the entire tree).

That said, it's not hard to implement a nested-bracket approach; I'll try to implement that too so people can experiment with both.

My main reason for having a slight preference for the explicit depth style is that I think `tr(KEY,[A|[[B|C]|D]])` is much harder to interpret by a human than `tr(KEY,1:A,3:B,3:C,2:D)`. In the latter it's immediately clear that B and C are the least likely paths, and A is the most likely one, for example. I agree with the argument that it's harder to write, but I'm not sure that's a big problem; for nontrivial things I suspect they'll either be written once as a template, or generated from a higher-level specification.

JeremyRubin commented:
I think it just depends on what descriptors are for v.s. what people end up using them for. Of that I am unsure, but it's worthwhile to note that this is a human readable format so presumably at some point a human may look at it or want to edit it.

w.r.t. annotations for "descriptor blinding", I actually think this is somewhat relevant. One could imagine a scenario where a branch is used both as a pre-image for a has revelation and as a script. In that case, you might not want to reveal the entire tree to someone as it might give them the ability to spend from that branch as well. 

w.r.t. brackets or list, it might be interesting to make the descriptors a NEWICK format subset, which seems to be the most commonly used tree format (and is a bracket based format). Then you could dump descriptors into standard tree viewing software easily. I couldn't find a good reference for tree file formats using a list/depth approach.

sipa commented:
@JeremyRubin @roconnor-blockstream Added a commit to switch to "tr(KEY,{S,{{S,S},S}})" notation. I prefer {} over [] because the latter is already in use for key origin information.

JeremyRubin commented:
Thanks!

Is there any reason not to use parens ()? NEWICK compatibility would be fun.

JeremyRubin commented on src/util/spanparsing.cpp:
```diff
@@ -34,11 +34,11 @@ Span<const char> Expr(Span<const char>& sp)
     auto it = sp.begin();
     while (it != sp.end()) {
-        if (*it == '(') {
+        if (*it == '(' || *it == '{') {
```
don't we need to trace what type we are in if it is a paren or a {? 

E.g., `(}` would pass here no?

JeremyRubin replied:
I see. 

bikeshed:
I think something like this could be a little clearer:

```c++
// This function finds the end of the current expression and returns pair(expr_span, rest_span)
// trailing comma is included in the expr_span, e.g.
// `<expr A>,<expr B>` would return `<expr A>,` `<expr B>` 
std::pair<Span<const char>, Span<const char>> Expr(const Span<const char>& sp)
{
    size_t left = 0;
    size_t right = 0;
    auto it = sp.begin();
    while (it != sp.end()) {
        switch (*it) {
        case '(':
        case '{':
            ++left;
            break;
        case '}':
        case ')':
            ++right;
            break;
        }
        bool balanced = right == left;
        // completed a Expr by matching nonzero left/right
        if (balanced && right > 0) break;
        // we have a balanced term as a comma separated entity
        if (balanced && *it == ',') break;
        ++it;
    }
    return std::make_pair(sp.first(it-sp.begin), sp.subspan(it-sp.begin()));
}
```

That said it seem peculiar to me that `,` is left on the string -- maybe preferable to trim it here?

sipa replied:
That's how the current span-based parsing works. I think discussions about improvements to that are interesting but out of scope for this PR.

JeremyRubin commented on src/script/descriptor.cpp:
```diff
@@ -1143,36 +1143,46 @@ std::unique_ptr<DescriptorImpl> ParseScript(uint32_t& key_exp_index, Span<const
+                    continue;
+                }
+                auto sarg = Expr(expr);
+                auto subdesc = ParseScript(key_exp_index, sarg, ParseScriptContext::P2TR, out, error);
```
you could directly emplace_back the subdesc and then nullcheck/return nullptr, rather than needing to std::move it.

sipa replied:
Done.

JeremyRubin commented on src/script/descriptor.cpp:
```diff
@@ -1143,36 +1143,46 @@ std::unique_ptr<DescriptorImpl> ParseScript(uint32_t& key_exp_index, Span<const
+                    }
+                    counts.pop_back();
+                }
+                if (counts.size()) {
```
Behavior Note:

I believe this would be more robust to go before the while loop -- it shouldn't come up, because we're in a binary tree maybe -- but it would be better to explicitly check:
```c++
if (counts.size() && counts.back() == false) 
```
or something like
```c++
if (counts.size() && counts.back() <= 2)
```

if we want to limit how many terms in a nest are allowed.

right now this invariant is implicitly checked by the order of the code and it is non-obvious to me.

sipa replied:
I've restructured things a bit, and added a bunch of comments. Does this look better?

JeremyRubin commented on src/script/descriptor.cpp:
```diff
@@ -829,6 +839,45 @@ class WSHDescriptor final : public DescriptorImpl
+        TaprootBuilder builder;
+        assert(m_depths.size() == scripts.size());
+        for (size_t pos = 0; pos < m_depths.size(); ++pos) {
+            builder.Add(m_depths[pos], scripts[pos], TAPROOT_LEAF_TAPSCRIPT);
```
given that this is just iterating in order, it seems like we should be able to directly feed the TaprootBuilder from the tr parsing code and do away with the intermediate vectors?

sipa replied:
No, unfortunately this happens at a different layers. TaprootBuilder can only be invoked with concrete keys; those don't exist yet at parsing time.

JeremyRubin replied:
I see :(

I guess I'll mark it as a personal todo (no need to upset a working internal representation) to brainstorm if there's a more obvious way to represent the tree than the ordered depth annotations (even though it's equivalent and unambiguous, I still find it unintuitive)

sipa replied:
I like linear data structures that don't need recursion to process, when unnecessary.

JeremyRubin replied:
Yeah, I agree with that 100%. What I was thinking was a vec<uint128> where each bit represents a left/right on the tree. That way you can do a de-duplicate check to make sure each node is only occupied once. Then the order of the vec itself is no longer significant; it should not need to be preserved because you can sort them by value (if I'm not wrong) to get the insertion order. Depth could be computed by looking at your nearest neighbor and computing the "divergence" (or by encoding an extra byte to make it explicit -- unclear what is best here). 

With a sorted vec, this also enables some nicesties such as using std:: binary searches for O(log(n)) lookup by tree position.

JeremyRubin replied:
@roconnor-blockstream may also like that this format should be:
1. concatenative, where concat has operations that are O(n)
    - merge sort then either deduplicate check or run through and add an appropriate offset to branches
 2.  "normalizable" (shifting things down to the smallest next number) and "prefixable" (adding a constant to all numbers to preserve structure.

sipa replied:
Why do the properties of this internal, temporary, data structure matter? It does what it is supposed to do. If we ever need other operations on it, something else can be considered.

JeremyRubin replied:
I find the code really hard to understand. Conceivably if we want people to review/audit it, a conceptually simpler representation & set of algorithms should be easier to understand? The above suggestion may or may not be it, but I haven't been able to fully work out how the algorithms this PR is currently using work even after staring at them for a bit.

the properties matter less.

sipa replied:
I'll try to add more comments.

sipa replied:
Added a big comment explaining the data structure in script/standard.h. WDYT?

sipa replied:
Marking this as resolved. Feel free to comment if the result still isn't clear.

fanquake commented on src/script/descriptor.cpp:
```diff
@@ -886,6 +936,13 @@ std::unique_ptr<PubkeyProvider> ParsePubkeyInner(uint32_t key_exp_index, const S
+                std::copy(data.begin(), data.end(), fullkey + 1);
+                pubkey.Set(std::begin(fullkey), std::end(fullkey));
+                if (pubkey.IsFullyValid()) {
+                    return MakeUnique<ConstPubkeyProvider>(key_exp_index, pubkey);
```
nit: in d5a0dce8fe40153b07ec6e882e8b1568b05e3905: we can use `std::make_unique` in new code.

sipa replied:
Done.

sipa commented:
Switched to the `tr(KEY,{{S1,S2},{{...)` style descriptors.

Sjors commented:
> Switched to the `tr(KEY,{{S1,S2},{{...)` style descriptors.

Can you clarify this, maybe with a simple example, in the PR description or in [descriptors.md](https://github.com/bitcoin/bitcoin/blob/master/doc/descriptors.md)?

sipa commented:
@Sjors Rebased, and

> Can you clarify this, maybe with a simple example, in the PR description or in descriptors.md?

done.

sipa commented on doc/descriptors.md:
```diff
@@ -80,12 +83,17 @@ Descriptors consist of several types of expressions. The top level expression is
 
+`TREE` expressions:
+- any `SCRIPT` expression
+- An open brace `{`, a `TREE` expression, a comma `,`, a `TREE` expression, and a closing brace `}`
```
No, `{A}` is not a valid TREE. `A` is.

So you'd use `tr(KEY,A)` but `tr(KEY,{A,B})`. There is nothing special about the top level.

achow101 commented on src/psbt.cpp:
```diff
@@ -240,37 +240,51 @@ bool SignPSBTInput(const SigningProvider& provider, PartiallySignedTransaction&
-            return false;
+    bool have_all_spent_outputs = true;
+    std::vector<CTxOut> utxos(tx.vin.size());
+    for (size_t idx = 0; idx < tx.vin.size(); ++idx) {
```
In 1d3cc29fb86b614b76b4d97a5adab7f903fc3020 "Construct and use PrecomputedTransactionData in PSBT signing"

We should avoid iterating all of the PSBT inputs again when we are signing each input. Instead we should compute the `PrecomputedTransactionData` in the caller and pass that into `SignPSBTInput`.

This can be done in `CWallet::FillPSBT` where we are iterating the inputs anyways to get their previous txs from the wallet. We could be constructing the `PrecomputedTransactionData` at that time and then passing it into `FIllPSBT` which then passes it to `SignPSBTInput`.

sipa replied:
Done.

sipa commented on src/script/descriptor.cpp:
```diff
@@ -817,6 +827,45 @@ class WSHDescriptor final : public DescriptorImpl
+            ret += strprintf(",%i:", m_depths[pos]);
+            std::string tmp;
+            if (!m_subdescriptor_args[pos]->ToStringHelper(arg, tmp, priv, normalized)) return false;
+            ret += std::move(tmp);
```
Nice catch. Fixed, and added tests for roundtrippability of the descriptors.

sipa commented on src/script/descriptor.cpp:
```diff
@@ -696,10 +696,20 @@ class RawDescriptor final : public DescriptorImpl
+    }
 public:
-    PKDescriptor(std::unique_ptr<PubkeyProvider> prov) : DescriptorImpl(Vector(std::move(prov)), "pk") {}
+    PKDescriptor(std::unique_ptr<PubkeyProvider> prov, bool xonly) : DescriptorImpl(Vector(std::move(prov)), "pk"), m_xonly(xonly) {}
```
Done.

sipa commented:
Rebased.

Sjors commented:
I created a descriptor like so, to test scriptpath spending: `tr(tpub1/0/*,pk(tprv2/0/*))`. This way the wallet won't have private keys for the keypath spend (`tpub1`), so it has to use the script `pk(tprv2/0*)`.

This seems to work on Signet, but the transaction does not enter the mempool. `testmempoolaccept` complains `min relay fee not met`, so perhaps there's a fee calculation bug? Once I bump the fee it goes through.

Here's the original raw transaction:
```
0200000000010114a5df5cf5084ea003c2f021c4e22846238a61c89894fa786e556395048ed40c0000000000fdffffff0292a0070000000000225120a287e30feacc24a25a9a3e689ad0107223cb247fb9a14880cddf3e62f1453a6420a1070000000000160014bf15b210d9f53e5c7b002fe904af042098706c3a03406c11b79f9a417f9702a6117e3bb170956ece09af6b584b2d92e1e0e5cfcad725250c91e29db77751af36970c0e5676622cfcf6fb1c81a316dfcd349b1ea59d412220d0586a7eba2e92a38f31ba8a7d82c05b9bb101bf3a649f205bd8e1df7c428b6eac21c07a35becbfeccf3ed5367ff117681497d040e6996eeae3dd21acb123e8525729d227b0000
```

sipa commented:
@Sjors That's expected, and I don't really know how to deal with this. The problem is that we determine the fee for spending an output by using the dummy signer, but the dummy signer can't know which private keys will and won't be available. Since tr spending cost differs based on what is available, this will give an inaccurate result (it'll always just use the key path).

A possibility is extended descriptors with markers to indicate that certain keys won't ever be available. It's good to point this out, but I think the scope will take us a bit too far for this PR.

Sjors commented:
This is definitely a new can of worms that will need some thought, but I agree it can wait (if documented).

The main point of having leaves is to be able to spend when the main keypath key material isn't available. But we currently have no way of indicating that.

I don't think descriptors are the right place for this. Let's say I have 3 hardware wallets. The keypath is a 3-of-3 musig, and the script(s) are 2-of-3. Normally I would sign with all 3 in order to save fees and preserve privacy. But after a minor boating accident I only have two. I wouldn't necessarily want to create a new wallet, so nothing about the descriptors changes.

The natural place to indicate which leaf to use is probably the send command (especially fun for the GUI...).

fjahr commented:
Concept ACK 🚀

Sjors commented:
All prerequisites have been merged! Needs rebase.

MarcoFalke commented on src/pubkey.h:
```diff
@@ -234,7 +240,14 @@ class XOnlyPubKey
     const unsigned char& operator[](int pos) const { return *(m_keydata.begin() + pos); }
     const unsigned char* data() const { return m_keydata.begin(); }
-    size_t size() const { return m_keydata.size(); }
+    static constexpr size_t size() { return uint256::size(); }
```
in commit b6f6c58ef7e64f699102beef063b3b9f91c71ecd:

Could use `decltype(m_keydata)::size()` for self-documenting code?

sipa replied:
Done.

fjahr commented on src/script/interpreter.cpp:
```diff
@@ -1431,8 +1431,8 @@ void PrecomputedTransactionData::Init(const T& txTo, std::vector<CTxOut>&& spent
-    bool uses_bip341_taproot = false;
+    bool uses_bip143_segwit = force;
+    bool uses_bip341_taproot = force;
     for (size_t inpos = 0; inpos < txTo.vin.size(); ++inpos) {
```
in 474536660d:

May as well skip this `for` loop altogether if `force` is true, if I see that correctly.

sipa replied:
Done.

sipa commented on doc/descriptors.md:
```diff
@@ -54,20 +55,22 @@ Output descriptors currently support:
-- `sortedmulti(k,KEY_1,KEY_2,...,KEY_n)` (anywhere): k-of-n multisig script with keys sorted lexicographically in the resulting script.
+- `multi(k,KEY_1,KEY_2,...,KEY_n)` (not inside `tr`): k-of-n multisig script.
+- `sortedmulti(k,KEY_1,KEY_2,...,KEY_n)` (not inside `tr`): k-of-n multisig script with keys sorted lexicographically in the resulting script.
+- `tr(XKEY)` or `tr(KEY,TREE)` (top level only): P2TR output with the specified key as internal key, and optionally a tree of script paths.
```
Fixed.

sipa commented:
Rebased, and addressed comments.

MarcoFalke commented:
```
 node0 stderr bitcoin-node: script/interpreter.cpp:1714: virtual bool GenericTransactionSignatureChecker<CMutableTransaction>::CheckSchnorrSignature(Span<const unsigned char>, Span<const unsigned char>, SigVersion, const ScriptExecutionData &, ScriptError *) const [T = CMutableTransaction]: Assertion `this->txdata' failed.

Sjors commented on src/script/standard.cpp:
```diff
@@ -371,3 +374,99 @@ CScript GetScriptForMultisig(int nRequired, const std::vector<CPubKey>& keys)
+{
+    assert((leaf_version & ~TAPROOT_LEAF_MASK) == 0);
+    if (!IsValid()) return *this;
+    /* Construct NodeInfo object with leaf hash and - if desired - leaf information. */
```
f86a68c0b1a6ed9e0c7e6f0ee81989a6129abe30: what do you mean by "if desired - leaf information"?

sipa replied:
That comment didn't make sense in the commit where it was in. I've moved it, and made it a bit clearer.

Sjors reviewed (commented):
Reviewed down to f86a68c, will continue later...

The refactoring of consensus code in 5ecaca05a299d3da9cdfdf8e9c6860500a73587d (`CheckTapTweak`) deserves its own commit (which also makes the diff slightly easier to read).

sipa committed:
Use HandleMissingData also in CheckSchnorrSignature

sipa commented on src/uint256.h:
```diff
@@ -75,7 +75,7 @@ class base_blob
     }
 
-    unsigned int size() const
+    static constexpr unsigned int size()
```
Done, added commit "Make XOnlyPubKey act like byte container".

sipa commented on src/script/standard.h:
```diff
@@ -142,11 +149,11 @@ struct WitnessUnknown
  *  A CTxDestination is the internal data type encoded in a bitcoin address
  */
-using CTxDestination = std::variant<CNoDestination, PKHash, ScriptHash, WitnessV0ScriptHash, WitnessV0KeyHash, WitnessUnknown>;
+using CTxDestination = std::variant<CNoDestination, PKHash, ScriptHash, WitnessV0ScriptHash, WitnessV0KeyHash, WitnessUnknown, WitnessV1Taproot>;
```
I didn't do that because the descriptor code's `GetOutputType` depended on the exact order of elements in this variant. I've addressed that in a new commit "Avoid dependence on CTxDestination index order", and then implemented your suggestion of placing WitnessUnknown last.

sipa commented on src/script/standard.cpp:
```diff
@@ -242,8 +242,13 @@ bool ExtractDestination(const CScript& scriptPubKey, CTxDestination& addressRet)
-    case TxoutType::WITNESS_UNKNOWN:
     case TxoutType::WITNESS_V1_TAPROOT: {
+        WitnessV1Taproot tap;
+        std::copy(vSolutions[1].begin(), vSolutions[1].end(), tap.begin());
```
The explanation is just that this was easier, because while `TxoutType` did have a separation between unknown and taproot, CTxDestination did not. Making them both produce {version, program} as `vSolutions` was easier because it allowed handling them identically in `ExtractDestination`.

I've addressed this by first changing the Solver output for taproot outputs to be just {program}, like P2WPKH/P2WSH, in a new commit "c1445ee0d56e56d824bc79a0c9f1db26105ade1f", and then used that in the later commits.

sipa commented:
Note: if this is too much, it'd be totally reasonable to split this PR in two, splitting after "Add tr() descriptor (derivation only, no signing)". Everything up to there is preparation & descriptor derivation. Everything after is signing logic.

Sjors commented:
Being able to make watch-only taproot wallet seems like a nice natural point to split this PR.

sipa commented:
I split off the derivation part in #22051.

sipa commented:
Rebased after #22051 merge.

sipa commented on src/test/script_standard_tests.cpp:
```diff
@@ -379,4 +381,70 @@ BOOST_AUTO_TEST_CASE(script_standard_GetScriptFor_)
+    BOOST_CHECK(builder.IsValid() && builder.IsComplete());
+    builder.Finalize(key_inner);
+    BOOST_CHECK(builder.IsValid() && builder.IsComplete());
+    BOOST_CHECK_EQUAL(EncodeDestination(builder.GetOutput()), "bc1pj6gaw944fy0xpmzzu45ugqde4rz7mqj5kj0tg8kmr5f0pjq8vnaqgynnge");
```
`wallet_taproot.py` already tests address derivation against the python implementation (which is also used by `feature_taproot.py`, so I think that's unnecessary.

achow101 commented on src/script/sign.cpp:
```diff
@@ -104,6 +152,85 @@ static bool CreateSig(const BaseSignatureCreator& creator, SignatureData& sigdat
 
+static bool CreateTaprootScriptSig(const BaseSignatureCreator& creator, SignatureData& sigdata, const SigningProvider& provider, std::vector<unsigned char>& sig_out, const XOnlyPubKey& pubkey, const uint256& leaf_hash, SigVersion sigversion)
+{
+    auto lookup_key = std::make_pair(pubkey, leaf_hash);
+    auto it = sigdata.taproot_script_sigs.find(lookup_key);
+    if (it != sigdata.taproot_script_sigs.end()) {
+        sig_out = it->second;
+    }
```
In ea07a01b0cf7256304b17508debf74d94cccdac7 "Basic Taproot signing logic in script/sign.cpp"

Can you explain what this is doing and why?

sipa replied:
It's the equivalent of `SignatureData::signatures`, but for Taproot/Schnorr signatures. They're not indexed by CKeyId, but by full (untweaked) pubkey and leaf they occur in.

This is redundant for now, as there is no multiparty signing whatsoever, and no way of communicating partial signatures (needs PSBT extension), but this is pretty much what it would look like. I could drop it for this PR.

achow101 commented:
Since #22166 uses "Add TaprootSpendData data structure, equivalent to script map for P2[W]SH" from here but with scripts as a multimap, in the interest of reducing future merge conflicts, I think this PR should use the same commit and use the multimap too. I expect that the change to using a multimap might cause some problems with the later commits in this PR.

sipa commented on src/script/standard.cpp:
```diff
@@ -386,6 +396,17 @@ bool IsValidDestination(const CTxDestination& dest) {
 
+void TaprootSpendData::Merge(TaprootSpendData other)
+{
+    if (internal_key.IsNull() && !other.internal_key.IsNull()) {
```
Asserting sounds dangerous, as I'd expect TaprootSpendData to at some point be untrusted data (e.g. gathered from PSBT). I agree it's a bit ugly to just prefer one over the other in case of a contradiction, and I also don't have a better solution. Would you be ok with just a TODO here?

sipa committed:
Make GetInputUTXO safer: verify non-witness UTXO match

sipa commented on src/script/standard.cpp:
```diff
@@ -386,6 +396,17 @@ bool IsValidDestination(const CTxDestination& dest) {
+{
+    if (internal_key.IsNull() && !other.internal_key.IsNull()) {
+        internal_key = other.internal_key;
+        merkle_root = other.merkle_root;
```
Done.

sipa commented on src/script/standard.h:
```diff
@@ -261,6 +284,7 @@ class TaprootBuilder
 
     XOnlyPubKey m_internal_key;  //!< The internal key, set when finalizing.
     XOnlyPubKey m_output_key; //!< The output key, computed when finalizing. */
+    bool m_parity;            //!< The tweak parity, computed when finalizing. */
```
Done. Also realigned.

sipa commented on src/test/key_tests.cpp:
```diff
@@ -302,4 +302,49 @@ BOOST_AUTO_TEST_CASE(bip340_test_vectors)
+        // Run the untweaked test vectors above, comparing with exact expected signature.
+        CKey key;
+        key.Set(sec.begin(), sec.end(), true);
+        XOnlyPubKey pubkey(key.GetPubKey());
```
Done.

sipa commented on src/psbt.cpp:
```diff
@@ -227,7 +227,41 @@ void UpdatePSBTOutput(const SigningProvider& provider, PartiallySignedTransactio
+        if (!input.witness_utxo.IsNull()) {
+            utxos[idx] = input.witness_utxo;
+            continue;
+        }
```
Changed to use `GetInputUTXO`. I've added a commit to add the prevout hash check to `GetInputUTXO` itself.

sipa commented:
Addressed most of your comments, and also updated the TaprootSpendData commit to use multiset semantics. Instead of using `multiset<key, control_block>`, I've changed it to `map<key, set<control_block>>`, which also has multiset semantics that should be compatible with #22166's inference code, but also with the signing logic in this PR (which should prefer the smallest control block for a given script if multiple are available).

sipa commented on src/psbt.cpp:
```diff
@@ -227,7 +227,41 @@ void UpdatePSBTOutput(const SigningProvider& provider, PartiallySignedTransactio
+    bool have_all_spent_outputs = true;
+    std::vector<CTxOut> utxos(tx.vin.size());
+    for (size_t idx = 0; idx < tx.vin.size(); ++idx) {
+        if (psbt.inputs.size() <= idx) {
```
Gone by rewriting.

achow101 commented:
ACK 4ad2a1d4adfcff715fed3ec37433ffd507c98b84

S3RK reviewed (commented):
I have a few simple questions, please help me understand.

sipa commented on src/script/descriptor.cpp:
```diff
@@ -847,7 +847,9 @@ class TRDescriptor final : public DescriptorImpl
         builder.Finalize(xpk);
-        return Vector(GetScriptForDestination(builder.GetOutput()));
+        WitnessV1Taproot output = builder.GetOutput();
+        out.tr_spenddata[output].Merge(builder.GetSpendData());
```
Well, not right now.

But like several things in this PR it's sort of written to accomodate future PSBT extensions that expose some of the TaprootSpendData. And one of the possbilities is only including information about a subset of leaves. In that case, you do want a merge operation when combining.

sipa commented on src/script/descriptor.cpp:
```diff
@@ -847,7 +847,9 @@ class TRDescriptor final : public DescriptorImpl
         XOnlyPubKey xpk(keys[0]);
         if (!xpk.IsFullyValid()) return {};
```
There is no real difference. An unspendable internal pubkey is just a pubkey with no known private key. You still need to actually select one.

michaelfolkson replied:
@S3RK: More details here https://bitcoin.stackexchange.com/questions/99722/taproot-eliminating-key-path

sipa commented on src/script/signingprovider.h:
```diff
@@ -50,11 +52,13 @@ struct FlatSigningProvider final : public SigningProvider
     std::map<CKeyID, CPubKey> pubkeys;
     std::map<CKeyID, std::pair<CPubKey, KeyOriginInfo>> origins;
     std::map<CKeyID, CKey> keys;
+    std::map<XOnlyPubKey, TaprootSpendData> tr_spenddata;
```
Done.

achow101 commented:
re-ACK c69ad7e8097fb6a6cfdd28f135e666ba40405b61

sipa commented:
Rebased.

achow101 commented:
re-ACK 458a345b0590fd2fa04c7d8d70beb8d57e34bbc8

jonatack commented:
Reviewing tonight/tomorrow.

Sjors commented:
I'll try to review this today. Having this feature in the signet / testnet wallet makes it easier for people to test taproot before it activates, and potentially find problems (bugs, many eyes, shallow, etc). So it seems worth merging even slightly after the feature freeze.

michaelfolkson commented:
Concept ACK, Approach ACK. Updated syntax choice makes sense to me under the readability assumption that use cases that need more than say 2 layers will be rarer. Ignoring readability, the updated syntax choice seems superior due to arguments given.

Agree with @Sjors on importance of getting this in for 22.0. Two weeks for any bug fixes from today should be ok and protections in place (e.g. https://github.com/bitcoin/bitcoin/pull/22156) regardless to prevent users from spending mainnet Taproot outputs pre-activation (November).

Sjors commented:
Very light testing using this PR rebased onto #22260: I'm still able to spend from my `tr(xpub...)` and `tr(xpub1, pk(xpub2))` signet descriptor wallets in the GUI.

fjahr commented on src/script/sign.cpp:
```diff
@@ -48,6 +51,51 @@ bool MutableTransactionSignatureCreator::CreateSig(const SigningProvider& provid
+
+    CKey key;
+    {
+        // For now, use the old full pubkey-based key derivation logic. As it indexed by
```
As it *is* indexed by

sipa replied:
Done in https://github.com/bitcoin/bitcoin/pull/22275.

fjahr reviewed (commented):
Code review ACK 458a345b0590fd2fa04c7d8d70beb8d57e34bbc8

I don't consider any of my comments blocking. Will test over the next days.

Sjors reviewed (approved):
ACK 458a345b0590fd2fa04c7d8d70beb8d57e34bbc8

S3RK commented on src/key.h:
```diff
@@ -128,6 +128,18 @@ class CKey
+     * optionally tweaked by *merkle_root. Additional nonce entropy can be provided through
+     * aux.
+     *
+     * When merkle_root is not nullptr, this results in a signature with a modified key as
```
What's the use case for non-tweaked signatures?

nit: The difference between `merkle_root == nullptr` and `merkel_root->IsNull()` is subtle enough that
one might be mistaken, especially when passing the `aux` parameter and not reading the comments closely. Could this be a potential problem in the future?

sipa replied:
Signatures with OP_CHECKSIG and friends aren't tweated; only key-path signatures are.

So we do need both a way to ask for untweaked, and for tweaked with nothing. I agree it's confusing, though we need a way to ask for both. Getting it wrong is exceedingly likely to cause validation failures, though.

Sjors replied:
Yikes, I missed that subtlety too. In #22275 you expand the comment bit, e.g.:
```
* When merkle_root is nullptr, as distinct from merkle_root->IsNull(), the signature is not tweaked.
* This is the case for e.g. OP_CHECKSIG in a tapscript. For key-path signature on the other hand the signature is always tweaked, as follows: 
```

sipa replied:
Added something like that to https://github.com/bitcoin/bitcoin/pull/22275.

S3RK reviewed (commented):
two comments

sipa commented on src/script/sign.cpp:
```diff
@@ -104,6 +152,86 @@ static bool CreateSig(const BaseSignatureCreator& creator, SignatureData& sigdat
+                sigdata.taproot_key_path_sig = sig;
+            }
+        }
+        if (sigdata.taproot_key_path_sig.size()) {
```
I tried this for a bit, but I don't find the result cleaner.

sipa commented on src/script/signingprovider.h:
```diff
@@ -50,11 +52,13 @@ struct FlatSigningProvider final : public SigningProvider
     std::map<CKeyID, CPubKey> pubkeys;
     std::map<CKeyID, std::pair<CPubKey, KeyOriginInfo>> origins;
     std::map<CKeyID, CKey> keys;
+    std::map<XOnlyPubKey, TaprootSpendData> tr_spenddata; /** Map from output key to spend data. */
```
It's a `struct`, not a `class`, so the developer notes don't require that (it's inconsistently applied through the codebase, in some cases using `m_` prefixes for struct members too, but I don't think that's enough reason to do so here).

sipa commented on src/key.h:
```diff
@@ -128,6 +128,18 @@ class CKey
+     * When merkle_root is not nullptr, this results in a signature with a modified key as
+     * specified in BIP341:
+     * - If merkle_root->IsNull(): key + H_TapTweak(pubkey)*G
+     * - Otherwise:                key + H_TapTweak(pubkey || *merkle_root)
```
It's referring to the operation applied to the private key, not the public one.

sipa replied:
Oh, so the the line before is wrong. I see. Added in https://github.com/bitcoin/bitcoin/pull/22275.

sipa commented on src/script/sign.h:
```diff
@@ -40,9 +40,11 @@ class MutableTransactionSignatureCreator : public BaseSignatureCreator {
 
 public:
     MutableTransactionSignatureCreator(const CMutableTransaction* txToIn, unsigned int nInIn, const CAmount& amountIn, int nHashTypeIn = SIGHASH_ALL);
+    MutableTransactionSignatureCreator(const CMutableTransaction* txToIn, unsigned int nInIn, const CAmount& amountIn, const PrecomputedTransactionData* txdata, int nHashTypeIn = SIGHASH_ALL);
```
Fixed in https://github.com/bitcoin/bitcoin/pull/22275; just removed the default argument value.

sipa commented on src/script/sign.cpp:
```diff
@@ -476,6 +633,26 @@ bool SignTransaction(CMutableTransaction& mtx, const SigningProvider* keystore,
+
+    PrecomputedTransactionData txdata;
+    std::vector<CTxOut> spent_outputs;
+    spent_outputs.resize(mtx.vin.size());
+    bool have_all_spent_outputs = true;
+    for (unsigned int i = 0; i < mtx.vin.size(); i++) {
+        CTxIn& txin = mtx.vin[i];
+        auto coin = coins.find(txin.prevout);
+        if (coin == coins.end() || coin->second.IsSpent()) {
+            have_all_spent_outputs = false;
+        } else {
+            spent_outputs[i] = CTxOut(coin->second.out.nValue, coin->second.out.scriptPubKey);
+        }
+    }
+    if (have_all_spent_outputs) {
+        txdata.Init(txConst, std::move(spent_outputs), true);
+    } else {
+        txdata.Init(txConst, {}, true);
+    }
```
Done in https://github.com/bitcoin/bitcoin/pull/22275.

sipa commented on src/script/interpreter.cpp:
```diff
@@ -1478,8 +1478,8 @@ PrecomputedTransactionData::PrecomputedTransactionData(const T& txTo)
 // explicit instantiation
-template void PrecomputedTransactionData::Init(const CTransaction& txTo, std::vector<CTxOut>&& spent_outputs);
-template void PrecomputedTransactionData::Init(const CMutableTransaction& txTo, std::vector<CTxOut>&& spent_outputs);
+template void PrecomputedTransactionData::Init(const CTransaction& txTo, std::vector<CTxOut>&& spent_outputs, bool force);
```
Done in https://github.com/bitcoin/bitcoin/pull/22275.

sipa commented on src/script/interpreter.cpp:
```diff
@@ -1431,9 +1431,9 @@ void PrecomputedTransactionData::Init(const T& txTo, std::vector<CTxOut>&& spent
-    for (size_t inpos = 0; inpos < txTo.vin.size(); ++inpos) {
+    bool uses_bip143_segwit = force;
+    bool uses_bip341_taproot = force;
+    for (size_t inpos = 0; inpos < txTo.vin.size() && !(uses_bip143_segwit && uses_bip341_taproot); ++inpos) {
```
Too late.

sipa commented on src/script/standard.h:
```diff
@@ -209,15 +210,38 @@ CScript GetScriptForRawPubKey(const CPubKey& pubkey);
         /** Merkle hash of this node. */
         uint256 hash;
+        /** Tracked leaves underneath this node (either from the node itself, or its children).
+         *  The merkle_branch field for each is the partners to get to *this* node. */
```
Done in https://github.com/bitcoin/bitcoin/pull/22275.

hebasto commented on src/script/standard.cpp:
```diff
@@ -386,6 +396,21 @@ bool IsValidDestination(const CTxDestination& dest) {
+        merkle_root = other.merkle_root;
+    }
+    for (auto& [key, control_blocks] : other.scripts) {
+        scripts[key].merge(std::move(control_blocks));
```
On macOS Mojave 10.14.6 (18G9216) system clang fails to compile this line due to the lack of `std::unordered_map::merge` support.

```
$ clang --version
Apple LLVM version 10.0.1 (clang-1001.0.46.4)
Target: x86_64-apple-darwin18.7.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
```

See: https://stackoverflow.com/questions/50444908/clang-6-not-supporting-unordered-mapmerge

Reported by @S3RK.

MarcoFalke replied:
I presume this would also make it harder to compile on https://packages.debian.org/stretch/gcc

hebasto replied:
Since #20413 we support gcc 7+ only, right?

MarcoFalke replied:
In theory there is https://packages.debian.org/stretch/clang-7, but I don't think anyone was able to compile master with system packages on stretch recently.

MarcoFalke replied:
Confirmed (for fun) that the issue exists with libc++7. 

```
# make V=1
Making all in src
make[1]: Entering directory '/bitcoin/src'
make[2]: Entering directory '/bitcoin/src'
make[3]: Entering directory '/bitcoin'
make[3]: Leaving directory '/bitcoin'
/usr/bin/ccache clang++-7 -stdlib=libc++ -std=c++17 -DHAVE_CONFIG_H -I. -I../src/config   -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=2 -I. -I./secp256k1/include  -DBOOST_SP_USE_STD_ATOMIC -DBOOST_AC_USE_STD_ATOMIC -I/usr/include -I./leveldb/include -I./leveldb/helpers/memenv -I./univalue/include -DHAVE_BUILD_INFO -D__STDC_FORMAT_MACROS -DPROVIDE_FUZZ_MAIN_FUNCTION -fdebug-prefix-map=/bitcoin/src=. -Wstack-protector -fstack-protector-all -Wall -Wextra -Wgnu -Wformat -Wformat-security -Wvla -Wshadow-field -Wswitch -Wthread-safety -Wrange-loop-analysis -Wredundant-decls -Wunused-variable -Wunused-member-function -Wdate-time -Wconditional-uninitialized -Wsign-compare -Woverloaded-virtual -Wunreachable-code-loop-increment -Wno-unused-parameter -Wno-self-assign -Wno-unused-local-typedef -Wno-implicit-fallthrough    -fPIE -g -O2 -MT script/libbitcoin_common_a-standard.o -MD -MP -MF script/.deps/libbitcoin_common_a-standard.Tpo -c -o script/libbitcoin_common_a-standard.o `test -f 'script/standard.cpp' || echo './'`script/standard.cpp
script/standard.cpp:410:22: error: no member named 'merge' in 'std::__1::set<std::__1::vector<unsigned char, std::__1::allocator<unsigned char> >, ShortestVectorFirstComparator, std::__1::allocator<std::__1::vector<unsigned char, std::__1::allocator<unsigned char> > > >'
        scripts[key].merge(std::move(control_blocks));
        ~~~~~~~~~~~~ ^
1 error generated.
Makefile:7756: recipe for target 'script/libbitcoin_common_a-standard.o' failed
make[2]: *** [script/libbitcoin_common_a-standard.o] Error 1
make[2]: Leaving directory '/bitcoin/src'
Makefile:15489: recipe for target 'all-recursive' failed
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory '/bitcoin/src'
Makefile:820: recipe for target 'all-recursive' failed
make: *** [all-recursive] Error 1
```

<!--

export DEBIAN_FRONTEND=noninteractive && apt update && apt install curl wget htop git vim ccache -y && git clone https://github.com/bitcoin/bitcoin.git && cd bitcoin && git checkout master && apt install build-essential libtool autotools-dev automake pkg-config bsdmainutils python3-zmq     libevent-dev libboost-system-dev libboost-filesystem-dev libboost-test-dev libboost-thread-dev  libsqlite3-dev  libdb++-dev  -y   &&  ./autogen.sh && apt install libqt5gui5 libqt5core5a libqt5dbus5 qttools5-dev qttools5-dev-tools -y && apt install clang-7 llvm-7 libc++abi-7-dev libc++-7-dev -y && ./configure CC=clang-7 CXX="clang++-7 -stdlib=libc++"  --with-incompatible-bdb && make -j $(nproc) src/bitcoind

hebasto replied:
See #22339.
